### PR TITLE
Affirm payload api endpoint

### PR DESCRIPTION
**Pull request by peterberkenbosch**
This PR will provide an endpoint for rendering the expected json data in the javascript block that will be posted to Affirm. 

- [x] Setup api endpoint controller and action
- [x] Setup the needed configuration for confirm and cancel urls
- [x] Render the correct json using JBuilder

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -1,12 +1,18 @@
 source 'https://rubygems.org'
 
-gem 'solidus', github: 'solidusio/solidus', branch: ENV.fetch('SOLIDUS_BRANCH', 'master')
+branch = ENV.fetch('SOLIDUS_BRANCH', 'master')
+gem 'solidus', github: 'solidusio/solidus', branch: branch
+
+if branch == 'master' || branch >= "v2.0"
```
Comment (Senjai) [inline Gemfile]: A check against `Rails::VERSION::MAJOR >= 5` here is better, since this has nothing to do with solidus, it's a raisl 5 change.

Additionally we might want to look at avoiding writing any new controller tests and use request specs instead. Not sure if company has committed to that approach, but that's probably a discussion that needs to happen.

File: app/controllers/spree/api/affirm_controller.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+module Spree
+  module Api
+    class AffirmController < Spree::Api::BaseController
+      def payload
```
Comment (Senjai) [inline app/controllers/spree/api/affirm_controller.rb]: This controller doesn't do anything in this commit, so it probably belongs in another commit.

File: config/routes.rb [hunk 3]
```diff
@@ -1,3 +1,7 @@
 Spree::Core::Engine.routes.draw do
-  # Add your extension routes here
+  namespace :api, defaults: { format: 'json' } do
+    resource :affirm, only: [], controller: 'affirm' do
```
Comment (Senjai) [inline config/routes.rb]: See previous comment about this route not doing anything

File: spec/controllers/spree/api/affirm_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,12 @@
+require 'spec_helper'
+
+module Spree
+  describe Api::AffirmController, type: :controller do
+    render_views
```
Comment (Senjai) [inline spec/controllers/spree/api/affirm_controller_spec.rb]: Should just be a request spec if we're going to rendre views

File: spec/controllers/spree/api/affirm_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+    describe "#payload" do
+      it "renders the correct Affirm json payload" do
+        get :payload
```
Comment (Senjai) [inline spec/controllers/spree/api/affirm_controller_spec.rb]: This doesn't test anything.

File: spec/spec_helper.rb [hunk 6]
```diff
@@ -30,6 +30,8 @@
 require 'spree/testing_support/factories'
 require 'spree/testing_support/url_helpers'
 
+require 'spree/api/testing_support/helpers'
```
Comment (Senjai) [inline spec/spec_helper.rb]: Do we use these helpers in our tests yet? I actually don't know.

File: app/controllers/spree/api/affirm_controller.rb [hunk 7]
```diff
@@ -1,7 +1,11 @@
 module Spree
   module Api
     class AffirmController < Spree::Api::BaseController
+      include Spree::Core::ControllerHelpers::Order
+      include Spree::Core::ControllerHelpers::Auth
```
Comment (Senjai) [inline app/controllers/spree/api/affirm_controller.rb]: If we're using an API controller, we should not be using the session for authorization. API's that use sessions are vulnerable to CSRF. Perhaps this should be a regular controller instead?

File: app/controllers/spree/api/affirm_controller.rb [hunk 8]
```diff
@@ -1,7 +1,11 @@
...
+
       def payload
+        @order = current_order
```
Comment (Senjai) [inline app/controllers/spree/api/affirm_controller.rb]: We don't use @order here for anything.

File: app/views/spree/api/affirm/payload.json.jbuilder [hunk 9]
```diff
@@ -0,0 +1,77 @@
+json.config do
```
Comment (Senjai) [inline app/views/spree/api/affirm/payload.json.jbuilder]: This commit should be in the same commits that define the endpoint that renders the view, and provides the tests for the endpoint.

Did we want to use JBuilder here? If we do, we enforce that dependency on everyone who uses the extension. If we went with a basic model based approach, either with AMS or a regular class, people can generate this json if they decide to use the extension in a way we have not originally accounted for.

File: config/routes.rb (hunk 3 above)
Comment (Senjai) [inline config/routes.rb]: What if people want to use their own controller at this route? What if they dont want the extension to be generating routes? Do we need to route in the spree namespace or can we define our own routes?

File: spec/controllers/spree/api/affirm_controller_spec.rb [hunk 10]
```diff
@@ -3,9 +3,15 @@
 module Spree
   describe Api::AffirmController, type: :controller do
     render_views
+
+    before do
+      stub_authentication!
```
Comment (Senjai) [inline spec/controllers/spree/api/affirm_controller_spec.rb]: This commit should be squashed with the commit that added the tests if we're just updating it to be more correct.

File: spec/controllers/spree/api/affirm_controller_spec.rb [hunk 11]
```diff
@@ -12,6 +12,7 @@ module Spree
       it "renders the correct Affirm json payload" do
         get :payload, format: :json
         expect(response).to be_success
+        expect(response.body).to_not be_empty
```
Comment (Senjai) [inline spec/controllers/spree/api/affirm_controller_spec.rb]: Squash

File: spec/controllers/spree/api/affirm_controller_spec.rb [hunk 12]
```diff
@@ -4,11 +4,17 @@ module Spree
   describe Api::AffirmController, type: :controller do
     render_views
 
+    let(:user) { create(:user) }
+    let(:current_api_user) { user }
+    let(:order) { create(:order_ready_to_complete, user: user) }
```
Comment (Senjai) [inline spec/controllers/spree/api/affirm_controller_spec.rb]: Squash

File: spec/controllers/spree/api/affirm_controller_spec.rb [hunk 13]
```diff
@@ -13,10 +13,16 @@ module Spree
     end
 
     describe "#payload" do
+      subject { get :payload, format: :json }
+
       before { allow(controller).to receive(:current_order) { order } }
 
+      it "will render the jbuilder json template" do
+        expect(subject).to render_template("api/affirm/payload")
```
Comment (Senjai) [inline spec/controllers/spree/api/affirm_controller_spec.rb]: Squash

File: spec/controllers/spree/api/affirm_controller_spec.rb [hunk 14]
```diff
@@ -21,7 +21,7 @@ module Spree
         expect(subject).to render_template("api/affirm/payload")
       end
 
-      it "renders the correct Affirm json payload" do
+      xit "renders the correct Affirm json payload" do
```
Comment (Senjai) [inline spec/controllers/spree/api/affirm_controller_spec.rb]: We don't commit disabled specs unless the specs were previously merged into master, and need to be temporarily disabled as a regression.

Because we haven't committed to master a disabled spec should not exist.

File: app/views/spree/api/affirm/payload.json.jbuilder [hunk 15]
```diff
@@ -51,8 +51,6 @@ json.items @order.line_items do |line_item|
   json.sku line_item.sku
   json.unit_price line_item.price.to_money.cents
   json.qty line_item.quantity
-  # This is shitty, the routing here is handled by angular
```
Comment (Senjai) [inline app/views/spree/api/affirm/payload.json.jbuilder]: Should be amended to the commit that added the payload.

File: Gemfile [hunk 16]
```diff
@@ -9,6 +9,7 @@ end
 
 gem 'pg'
 gem 'mysql2'
+gem 'jbuilder', github: 'rails/jbuilder'
```
Comment (Senjai) [inline Gemfile]: Should be amended to the commit that added the dependency

File: config/routes.rb [hunk 17]
```diff
@@ -4,4 +4,7 @@
       get :payload
     end
   end
+
+  post '/affirm/confirm', to: "affirm#confirm", as: :confirm_affirm
```
Comment (Senjai) [inline config/routes.rb]: These routes don't point anywhere as of this commit. They should be included in the commit that impliments their functionality, ideally one by one.

Comment (Senjai) [inline config/routes.rb]: they should probably also be included inside the resource block.

File: app/views/spree/api/affirm/payload.json.jbuilder [hunk 18]
```diff
@@ -1,6 +1,6 @@
-json.config do
-  json.financial_product_key Spree::Config.get(:affirm_financial_product_key)
-end
+# json.config do
```
Comment (Senjai) [inline app/views/spree/api/affirm/payload.json.jbuilder]: We dont commit commented out code. We commit comments only when they add meaningful context to the code.

File: app/views/spree/api/affirm/payload.json.jbuilder [hunk 19]
```diff
@@ -51,7 +51,7 @@ json.items @order.line_items do |line_item|
   json.sku line_item.sku
   json.unit_price line_item.price.to_money.cents
   json.qty line_item.quantity
-  json.item_url spree_product_url(line_item.product)
+  json.item_url spree.product_url(line_item.product)
```
Comment (Senjai) [inline app/views/spree/api/affirm/payload.json.jbuilder]: Should be amended to the commit that added this line.

File: spec/controllers/spree/api/affirm_controller_spec.rb [hunk 20]
```diff
@@ -21,7 +21,7 @@ module Spree
         expect(subject).to render_template("api/affirm/payload")
       end
 
-      xit "renders the correct Affirm json payload" do
+      it "renders the correct Affirm json payload" do
```
Comment (Senjai) [inline spec/controllers/spree/api/affirm_controller_spec.rb]: Squash

File: solidus_affirm.gemspec [hunk 21]
```diff
@@ -18,7 +18,7 @@ Gem::Specification.new do |s|
   s.files = Dir["{app,config,db,lib}/**/*", 'LICENSE', 'Rakefile', 'README.md']
   s.test_files = Dir['test/**/*']
 
-  s.add_dependency 'solidus', ['>= 1.1', '< 3']
+  s.add_dependency 'solidus', ['>= 1.4', '< 3']
```
Comment (Senjai) [inline solidus_affirm.gemspec]: Why was this decision made? The commit message doesn't indicate why this had to be done.

File: spec/controllers/spree/api/affirm_controller_spec.rb [hunk 22]
```diff
@@ -24,7 +24,18 @@ module Spree
       it "renders the correct Affirm json payload" do
         subject
         expect(response).to be_success
-        expect(response.body).to_not be_empty
+        expect(json_response).to_not be_empty
+
```
Comment (Senjai) [inline spec/controllers/spree/api/affirm_controller_spec.rb]: Squash.

File: app/views/spree/api/affirm/payload.json.jbuilder [hunk 23]
```diff
@@ -1,7 +1,3 @@
-# json.config do
-#   json.financial_product_key Spree::Config.get(:affirm_financial_product_key)
```
Comment (Senjai) [inline app/views/spree/api/affirm/payload.json.jbuilder]: squash

File: app/views/spree/api/affirm/payload.json.jbuilder [hunk 24]
```diff
@@ -6,38 +6,12 @@ end
 
 json.shipping do
   address = @order.ship_address
-  json.name do
-    json.first address.first_name
-    json.last address.last_name
-  end
-
-  json.address do
```
Comment (Senjai) [inline app/views/spree/api/affirm/payload.json.jbuilder]: Squash

File: app/views/spree/api/affirm/payload.json.jbuilder [hunk 25]
```diff
@@ -21,8 +21,10 @@ json.items @order.line_items do |line_item|
   json.qty line_item.quantity
   json.item_url spree.product_url(line_item.product)
 
-  if image = line_item.variant.images.first
-    json.item_image_url root_url + image.attachment.url
+  if item.variant.images.any?
```
Comment (Senjai) [inline app/views/spree/api/affirm/payload.json.jbuilder]: Squash

File: app/views/spree/api/affirm/payload.json.jbuilder [hunk 26]
```diff
@@ -21,10 +21,10 @@ json.items @order.line_items do |line_item|
   json.qty line_item.quantity
   json.item_url spree.product_url(line_item.product)
 
-  if item.variant.images.any?
-    json.item_image_url URI.join(root_url, item.variant.images.first.attachment.url(:large))
-  elsif item.variant.product.images.any?
-    json.item_image_url URI.join(root_url, item.variant.product.images.first.attachment.url(:large))
+  if line_item.variant.images.any?
```
Comment (Senjai) [inline app/views/spree/api/affirm/payload.json.jbuilder]: Squash

File: spec/controllers/spree/api/affirm_controller_spec.rb [hunk 27]
```diff
@@ -0,0 +1,12 @@
+require 'spec_helper'
+
+module Spree
```
Comment (Senjai) [inline spec/controllers/spree/api/affirm_controller_spec.rb]: prefer `RSpec.describe Spree::Api::AffirmController` instead

File: app/controllers/spree/api/affirm_controller.rb (hunk 7 above)
Comment (peterberkenbosch) [inline app/controllers/spree/api/affirm_controller.rb]: mmm good catch! Investigating this.

File: app/controllers/spree/api/affirm_controller.rb (hunk 8 above)
Comment (peterberkenbosch) [inline app/controllers/spree/api/affirm_controller.rb]: it's for the jbuilder view template.

File: config/routes.rb (hunk 3 above)
Comment (peterberkenbosch) [inline config/routes.rb]: - I see no good reason why people would like a different controller here yet
- There are routes needed to get the api endpoit
- the namespace question is a good one, might be better to define our own namespace for this extension to keep potential conflicts away.

File: config/routes.rb (hunk 17 above)
Comment (peterberkenbosch) [inline config/routes.rb]: yes, these should not be here at all!

File: spec/controllers/spree/api/affirm_controller_spec.rb (hunk 4 above)
Comment (peterberkenbosch) [inline spec/controllers/spree/api/affirm_controller_spec.rb]: yup

File: spec/spec_helper.rb (hunk 6 above)
Comment (peterberkenbosch) [inline spec/spec_helper.rb]: the stub_authentication is from there IIRC

File: app/views/spree/api/affirm/payload.json.jbuilder (hunk 9 above)
Comment (peterberkenbosch) [inline app/views/spree/api/affirm/payload.json.jbuilder]: would be the same case for AMS, (I do love AMS though :) ). I like the PORO approach

**Comment by peterberkenbosch** [comment]
Closing this in favor for #4
